VSC yellow paper (#382)
aeternity/protocol: channels/VSC.md

With this in mind, in order for us to explain in detail the nuts and bolts of
Virtual State Channels we must first dive deep into on-chain State Channels. I
will focus on the tricky parts regarding Virtual State Channels.

#### Alice and Bob

Let's have an Alice and a Bob. They open a State Channel on-chain using a
`channel_create` transaction. Then they wait a certain amount of confirmations
to be sure the opening transaction is safely burried deep enough in the chain
and only then do they proceed with using the channel.

```
  +-------------+      off-chain      +-------------+
  |   Alice     |+-------------------+|     Bob     |
  |-------------|       prtocol       |-------------|
  |             |                     |             |
  |  off-chain  |                     |  off-chain  |
  |    state    |                     |    state    |
  |    trees    |                     |    trees    |
  |             |                     |             |
  +-------------+                     +-------------+
```

Each channel has an ever changing state. It is represented by a set of MPTrees
that hold accounts, contracts and contract calls. This set of MPtrees can be

> treated as one big MPTree that has a root hash. This is what we call a

Author: yes

and was given time to dispute the solo close. There will be an uncertainty
margin when close to the `locked_until` height that cannot be completely
eliminated.

### Example scenarios

In order to provide better understanding of the proposition above, we will go
through a few hypotetical scenarios when something goes bad. We will answer
_What shall X do in order to be safe?_ for each one of them

#### Alice is malicious

##### Alice provides Ingrid with an old state

At any point of time Alice can provide Ingrid with some Virtual State Channel
State that might not be the latest. Ingrid then uses it to update her State
Channel with Bob. In this case Bob can simply provide a newer state and they all
are safe. This is important esp. while resolving a solo closing using Ingrid as
an arbiter.

##### Alice refuses to cooperate with Ingrid

If Ingrid has a newer state than the one Alice had provided, or Ingrid has a
state produced by a valid force progress by Bob, Alice might decide not to
cooperate. In this case Ingrid simply brings the issue on-chain using a forced
progress.

##### Alice produces a force progress based on an old state

> Ingrid receives a forced progress that seemsreasonable because the state it is

Reviewer: ```suggestion
Ingrid receives a forced progress that seems reasonable because the state it is
```

> based on a is the latest she had seen so far. Ingrid accepts it as valid and

Reviewer: ```suggestion
based on is the latest she had seen so far. Ingrid accepts it as valid and
```

would look similar like this:

```javascript=
{  
   "id": "ch_abc...", // the channel ID
   "initiator_id": "ak_bca...", // the address of the initiator
   "responder_id": "ak_123...", // the address of the responder
   "initiator_amount": 60, // the amount the initiator initally had locked in the channel
   "responder_amount": 40, // the amount the responder initally had locked in the channel
   "channel_amount": 100, // the sum of the two amounts
   "channel_reserve": 1, // a value provided by participants in channel_create transaction
   "delegate_ids": [], // for this example we don't need delegates
   "state_hash": "st_456...", // the state_hash from channel_create transaction
   "round": 1, // this is the initial value
   "solo_round": 0, // this is the initial value
   "lock_period": 0, // this is the initial value
   "locked_until": 0 // this is the initial value
}
```

Note the amounts as we will revisit those with regard of Virtual State Channels

#### Disputes

The general assumption is if the protocol is clean and easy to use, this would
provide incentive for participants to follow it. Cheating would be meaningless

> as it would yeld no benefit for the cheating party. That is a strong argument in

Reviewer: yield (yeld), typo

we used to sign only off-chain transactions.

Then Alice provides this off-chain update to her SC with Ingrid. This update
creates the Alice-Bob Virtual Channel object in their state trees. Alice
dedicates to it from her off-chain balance the 10 AE she had already agreed to
open a VSC with Bob. Ingrid on the other hand dedicates the 12 AE that Bob had
promised. Since we want the channel object to look as expected, the wrapper
object reflects the balance contributed by 'requester' and 'intermediary'
respectively. The channel object looks like this:

```javascript=
{
   "intermediary_id": "ak_...", // which one is Ingrid
   "intermediary_amount": 12, // part of the total contributed by Ingrid
   "requester_amount": 10, // part of the total contributed by Alice
   "channel" :
      {  
         "id": "ch_...", // the Alice-Bob channel ID
         "initiator_id": "ak_...", // the address of Alice
         "responder_id": "ak_...", // the address of Bob
         "initiator_amount": 10, // Alice balance
         "responder_amount": 12, // Bob balance
         "channel_amount": 22 // the sum of the two amounts
      } ...
}
```

> We could have an implicit logic that in the context of this particulat State

Reviewer: typo: particulat

#### Making an off-chain update in a Virtual State Channel

Once the channel has been opened, participants have a direct connection between
each other and can produce off-chain updates and off-chain states as they would
in an on-chain State Channel. The only difference is their authentication method
is according to their on-chain channel with Ingrid.

##### Example

Alice has an open VSC with Bob. Alice wants to execute a smart contract in their
VSC, Alice produces the updates she desires and prepares a corresponding
off-chain transaction. It has the AB Virtual State Channel ID, the next `round`
and the newly calculated `state_hash`. Alice authenticates this transaction
according to her authentication method used in her on-chain State Channel with
Ingrid. She then gives it to Bob to co-authenticate. Bob does so according to
his authentication method in his on-chain State Channel with Ingrid. He then
returns the co-authenticated transaction to Alice and the update is finished.

It is worth noting that this closly resembles the on-chain State Channel
protocol. Also it is important to stress that none of the off-chain updates
reach Ingrid.

> #### Depositing more tokens to a Virtual State Channel/withdrawing from a

Reviewer: ```suggestion
#### Depositing more tokens to a Virtual State Channel/withdrawing from a Virtual State Channel
```

> Virtual State Channel

Reviewer: ```suggestion
```

##### Virtual State Channel Force Progress

In order to improve privacy and not to provide the whole SC state, we define a
new transaction type for Virtual State Channel Force Progressing. Instead of
providing the whole SC state tree, only a Proof of Inclusion containing the
relevant object is provided. Depending on the specific progress would include
the channel object and optionally accounts to which tokens can be withdrawn
to.

The transaction is called `channel_vsc_force_progress` and it consists of:
* `payload` that defines the `round` and the `state_hash` the progress is base
  on. This could be:
    * the latest on-chain persisted one: in this case the on-chain stored
      `round` and `state_hash` are used. The `payload` is empty
    * off-chain transaction with a greater `round` to the one currently
      persisted on-chain for this SC. The `payload` is a serializated
      off-chain transaction and its `round` and `state_hash` are used for
      defining the SC state that the progress is being based upon.
* `poi` is a Proof of Inclusion that provides enough information for the
  forced progress to be applied. This would include the VSC object but

>   possibly accounts to which tockens are withdrawn to might be needed as well.

Reviewer 2: ```suggestion
  possibly accounts to which tokens are withdrawn might be needed as well.
```

* `update` is a VSC force progress update. This could be:
  * a co-authenticated VSC withdraw
  * a co-authenticated VSC snapshot
  * a co-authenticated VSC settle
  * a co-authenticated VSC mutual close

>   * unilaterally authenticated dispute. Those also provide a context of the

Reviewer 2: _a_ unilaterally ...
context _for_ ...

    execution: a `payload` that is either empty or a co-authenticated VSC
    off-chain state, VSC poi or trees, next VSC `state_hash` and next VSC
    `round`. Update types for disputes supported are:
    * solo close
    * slash
    * force progress
    * virtual force progress
* `state_hash` - the root of the SC state tree after the `update` had been
  applied on top of the `poi`
* `round`: SC's next round

> Note that this allows force progressing of both contracts in VSC and force

Reviewer 2: force-progressing

progressing a VSC inside a VSC.

From now on we will use force progress in the context of VSC meaning a Virtual

> State Channel Force progress transaction - `channel_vsc_force_progress`.

Reviewer 2: Force Progress

##### Example

Alice wants to close her Virtual State Channel with Bob, but Bob refuses. Alice
produces a solo update in her on-chain State Channel with Ingrid initiating a
solo-closing sequence of the Virtual State Channel. If Ingrid refuses this,

> Alice can force it on-chain using `channel_vsc_force_progress.

Reviewer 2: missing ending backtick.

#### Bob is not responding or missing

If there is a dispute between Alice and Bob or for the sake of the example Bob
refuses to co-authenticate the next state, Alice can produce an off-chain

> `channel_vsc_force_progress` which is similar to force progress and bring it

Reviewer 2: ..., which is similar to force progress, ... (commas)
